lib/phase/flatten.js

```diff
@@ -19,41 +42,91 @@ function flattenSharedIntoJobs(shared, jobs) {
         const oldJob = clone(jobs[jobName]);
 
         // Replace
-        ['image', 'matrix', 'steps'].forEach((key) => {
+        ['image', 'matrix', 'steps', 'template'].forEach((key) => {
             if (oldJob[key]) {
                 newJob[key] = oldJob[key];
             }
         });
 
-        if (!newJob.environment) {
-            newJob.environment = {};
-        }
-        if (!newJob.secrets) {
-            newJob.secrets = [];
-        }
-        if (!newJob.settings) {
-            newJob.settings = {};
-        }
-
-        // Merge
-        Object.assign(newJob.environment, oldJob.environment || {});
-        Object.assign(newJob.settings, oldJob.settings || {});
-        newJob.secrets = newJob.secrets.concat(oldJob.secrets || []);
-
+        merge(newJob, oldJob);
         newJobs[jobName] = newJob;
     });
 
     return newJobs;
 }
 
 /**
+ * Retrieve template and merge into job config
+ *
+ * @method mergeTemplateIntoJob
+ * @param  {String}           jobName           Job name
+ * @param  {Object}           jobConfig         Job config
+ * @param  {Object}           newJobs           List of new jobs
```

Comment:
Objects with jobs as key. I'm using the same comments as here: https://github.com/screwdriver-cd/config-parser/blob/master/lib/phase/flatten.js#L11

I can make it more clear.